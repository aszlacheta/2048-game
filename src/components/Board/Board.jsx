import { useEffect, useState } from "react";
import { getCells } from "../../api";
import Hexagon from "../Hexagon/Hexagon";
import * as R from 'ramda';

import './Board.scss';
import useKeyUp from "../../hooks/useKeyUp";
import GameStatus, { GAME_STATUS } from "../GameStatus/GameStatus";

export default function Board() {

    const EMPTY_BOARD = [
        { x: 0, y: 1, z: -1, value: 0 },
        { x: -1, y: 1, z: 0, value: 0 },
        { x: 1, y: 0, z: -1, value: 0 },
        { x: 0, y: 0, z: 0, value: 0 },
        { x: -1, y: 0, z: 1, value: 0 },
        { x: 1, y: -1, z: 0, value: 0 },
        { x: 0, y: -1, z: 1, value: 0 },
    ];

    const [cells, setCells] = useState(EMPTY_BOARD);
    const [status, setStatus] = useState(GAME_STATUS.gameOver);
    const [possibleMoves, setPossibleMoves] = useState(0);

    useEffect(() => {
        clearBoard();
        addNewCells();
        // eslint-disable-next-line
    }, [])

    const addNewCells = () => {
        getCells(getExistingCells(cells)).then(({ data }) => {
            update(data);
        });
    }

    useEffect(useKeyUp('w', () => shiftAndUpdate(() => shiftNorth(cells))), [cells, status]);
    useEffect(useKeyUp('s', () => shiftAndUpdate(() => shiftSouth(cells))), [cells, status]);
    useEffect(useKeyUp('q', () => shiftAndUpdate(() => shiftNorthWest(cells))), [cells, status]);
    useEffect(useKeyUp('a', () => shiftAndUpdate(() => shiftSouthWest(cells))), [cells, status]);
    useEffect(useKeyUp('e', () => shiftAndUpdate(() => shiftNorthEast(cells))), [cells, status]);
    useEffect(useKeyUp('d', () => shiftAndUpdate(() => shiftSouthEast(cells))), [cells, status]);

    useEffect(() => {
        const status = checkStatus(cells);
        setStatus(status);
    }, [cells]);

    const getExistingCells = (cells) => cells.filter(cell => cell.value !== 0) || [];

    const clearBoard = () => {
        setCells(EMPTY_BOARD);
    }

    const shiftAndUpdate = (callback) => {
        const data = callback();

        if (status === GAME_STATUS.playing) {
            update(data);
            addNewCells();
        }
    }

    const update = (data, updateData = true) => {
        const shallowCopyCells = [...cells];

        data.forEach(({ x, y, z, value }) => {
            const pred = R.whereEq({ x, y, z })
            const index = R.findIndex(pred)(cells);

            if (index >= 0) {
                shallowCopyCells[index] = { x, y, z, value };
            }
        });

        if (updateData) {
            setCells(shallowCopyCells);
        }

        return shallowCopyCells;
    }

    const shiftNorth = (cells) => {
        const groupByX = R.groupBy(cell => cell?.x);
        const groupedByX = groupByX(cells);

        shift(groupedByX);

        return Object.values(groupedByX).flat();
    }

    const shiftSouth = (cells) => {
        const groupByX = R.groupBy(cell => cell?.x);
        const groupedByX = groupByX(cells);

        reverseArraysInObjects(groupedByX);
        shift(groupedByX);
        reverseArraysInObjects(groupedByX);

        return Object.values(groupedByX).flat();
    }

    const shiftNorthWest = (cells) => {
        const groupByZ = R.groupBy(cell => cell?.z);
        const groupedByZ = groupByZ(cells);

        shift(groupedByZ);

        return Object.values(groupedByZ).flat();
    }

    const shiftSouthWest = (cells) => {
        const groupByY = R.groupBy(cell => cell?.y);
        const groupedByY = groupByY(cells);

        reverseArraysInObjects(groupedByY);
        shift(groupedByY);
        reverseArraysInObjects(groupedByY);

        return Object.values(groupedByY).flat();
    }

    const shiftNorthEast = (cells) => {
        const groupByY = R.groupBy(cell => cell?.y);
        const groupedByY = groupByY(cells);

        shift(groupedByY);

        return Object.values(groupedByY).flat();
    }

    const shiftSouthEast = (cells) => {
        const groupByZ = R.groupBy(cell => cell?.z);
        const groupedByZ = groupByZ(cells);

        reverseArraysInObjects(groupedByZ);
        shift(groupedByZ);
        reverseArraysInObjects(groupedByZ);

        return Object.values(groupedByZ).flat();
    }

    const shift = (groupedBy) => {
        Object.keys(groupedBy).forEach(x => {
            let rowByX = groupedBy[x];

            // filter zeros
            rowByX = filterZero(rowByX);

            // merge
            rowByX.forEach((cell, index) => {
                if (index + 1 < rowByX.length && cell.value === rowByX[index + 1].value) {
                    rowByX[index].value = 2 * cell.value;
                    rowByX[index + 1].value = 0;
                }
            });

            // filter zeros again
            rowByX = filterZero(rowByX);

            // update values
            groupedBy[x].forEach((cell, index) => {
                groupedBy[x][index].value = rowByX[index]?.value || 0;
            })
        })

        return groupedBy;
    }

    const filterZero = (array) => array.filter(cell => cell?.value !== 0);

    const reverseArraysInObjects = (array) => {
        Object.keys(array).forEach(x => {
            array[x].reverse();
        });
        return array;
    }

    const checkStatus = (cells) => {
        //TODO
        let shallowCopyCells = R.clone(cells);
        let possibleMoves = 0;

        let data = shiftNorth(shallowCopyCells);
        let a = R.difference(cells, data);
        if (a?.length > 0) {
            possibleMoves++;
        }

        shallowCopyCells = R.clone(cells);
        data = shiftNorthEast(shallowCopyCells);
        a = R.difference(cells, data);
        if (a.length > 0) {
            possibleMoves++;
        }


        shallowCopyCells = R.clone(cells);
        data = shiftNorthWest(shallowCopyCells);
        a = R.difference(cells, data);
        if (a.length > 0) {
            possibleMoves++;
        }


        shallowCopyCells = R.clone(cells);
        data = shiftSouth(shallowCopyCells);
        a = R.difference(cells, data);
        if (a.length > 0) {
            possibleMoves++;
        }


        shallowCopyCells = R.clone(cells);
        data = shiftSouthEast(shallowCopyCells);
        a = R.difference(cells, data);
        if (a.length > 0) {
            possibleMoves++;
        }


        shallowCopyCells = R.clone(cells);
        data = shiftSouthWest(shallowCopyCells);
        a = R.difference(cells, data);
        if (a.length > 0) {
            possibleMoves++;
        }

        if (possibleMoves > 0) {
            setPossibleMoves(possibleMoves);
            return GAME_STATUS.playing;
        } else

        setPossibleMoves(0);
            return GAME_STATUS.gameOver;
    }

    return (
        <div className="container">

            <GameStatus status={status} />
            <div>Possible moves {possibleMoves}</div>

            <div className="board">
                {cells.map((cell, index) =>
                    <Hexagon
                        key={index}
                        value={cell?.value}
                        x={cell?.x}
                        y={cell?.y}
                        z={cell?.z}
                    />)}
            </div>
        </div>
    )
}