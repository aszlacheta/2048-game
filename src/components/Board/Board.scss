.container {
    display: flex;
    height: 100%;
    justify-content: center;
    align-items: center;

    .statuses {
        display: flex;
        flex-direction: column;
        margin-right: 5rem;
    }

    .board {
        font-size: 0;
        width: var(--width);
        min-width: var(--width);
        height: calc(var(--width) * 0.9);
        --width: calc(var(--s) * var(--e));
        --e: 3;
        --m: 1px; // margin
        --f: calc(var(--s) + 4 * var(--m) - 1px); // used to shift some cells from the left

        &::before {
            content: "";
            width: calc(var(--s) / 1.25 - var(--ms) * var(--m));
            float: left;
            height: 150%;
            // background: repeating-linear-gradient(#000 0 10px, #0000 15px var(--f));
            shape-outside: repeating-linear-gradient(#000 0 10px, #0000 15px var(--f));
        }

        &[data-level="2"] {
            --e: 3; // max number of elements in row
            --s: 170px; // size
            --ms: 4;
            .hexagon {
                margin-right: calc(var(--s) / 1.75 - 4 * var(--m));

                &:nth-child(3n) {
                    margin-right: 0;
                }
            }
        }

        &[data-level="3"] {
            --e: 5; // max number of elements in row
            --s: 90px; // size
            --ms: 2;

            &::before {
                margin-top: calc(var(--s) * 1.5);
                height: calc(150% - var(--s) * 1.5);
            }

            .hexagon {
                font-size: 2.5rem;
                margin-right: calc(var(--s) * 0.5 + 4 * var(--m));
            }

            .hexagon:nth-child(1),
            .hexagon:last-child {
                margin-left: calc(var(--s) * 1.55 + 2 * var(--m));
            }

            .hexagon:nth-child(2) {
                margin-left: calc(var(--s) * 0.8);
            }
        }

        &[data-level="4"] {
            --e: 7; // max number of elements in row
            --s: 70px; // size
            --ms: 1;

            &::before {
                margin-top: calc(var(--s) * 2);
                height: calc(161% - var(--s) * var(--e));
            }

            .hexagon {
                font-size: 2rem;
                margin-right: calc(var(--s) * 0.5 + 4 * var(--m));
            }

            .hexagon:nth-child(1),
            .hexagon:last-child {
                margin-left: calc((var(--width) / 2) - (var(--s) + var(--e) * var(--m) + 3px) );
                margin-right: calc(var(--width) / 2 - var(--s) );
            }

            .hexagon:nth-child(2),
            .hexagon:nth-last-child(-n+3):not(:nth-last-child(-n+2)) {
                margin-left: calc(var(--s) * 1.6 - var(--ms) * var(--m) );
            }

            .hexagon:nth-child(4) {
                margin-left: calc(var(--s) * 0.75 + var(--e) * var(--ms) * var(--m) / 2 );
            }
        }
    }
}
