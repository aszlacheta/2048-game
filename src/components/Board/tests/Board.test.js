import renderer from 'react-test-renderer';
import Board from '../Board';

jest.mock("../../GameStatus/GameStatus");
jest.mock("../../PossibleMoves/PossibleMoves");
jest.mock("../../Hexagon/Hexagon");

describe('Board component', () => {

    const defaultsEmpty = {
        board: [],
        level: 2,
    }

    const defaults = {
        board: [
            { x: 0, y: 1, z: -1, value: 0 },
            { x: -1, y: 1, z: 0, value: 0 },
            { x: 1, y: 0, z: -1, value: 0 },
        ],
        level: 2,
    }

    it('should render properly empty board', () => {
        const component = renderer.create(<Board defaults={defaultsEmpty} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render properly non-empty board', () => {
        const component = renderer.create(<Board defaults={defaults} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    
});