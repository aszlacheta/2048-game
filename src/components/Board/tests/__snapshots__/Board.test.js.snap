// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component should render properly empty board 1`] = `
<div
  className="container"
>
  <div
    className="statuses"
  >
    <div
      className="mocked-game-status"
    />
    <div
      className="mocked-possible-moves"
    />
  </div>
  <div
    className="board"
  />
</div>
`;

exports[`Board component should render properly non-empty board 1`] = `
<div
  className="container"
>
  <div
    className="statuses"
  >
    <div
      className="mocked-game-status"
    />
    <div
      className="mocked-possible-moves"
    />
  </div>
  <div
    className="board"
  >
    <div
      className="mocked-hexagon"
    />
    <div
      className="mocked-hexagon"
    />
    <div
      className="mocked-hexagon"
    />
  </div>
</div>
`;
